modules/effects/arrow-effects/src/main/kotlin/arrow/effects/typeclasses/Bracket.kt

```diff
@@ -0,0 +1,83 @@
+package arrow.effects.typeclasses
+
+import arrow.Kind
+import arrow.core.Either
+import arrow.effects.typeclasses.ExitCase.Error
+import arrow.typeclasses.MonadError
+
+sealed class ExitCase<out E> {
+
+  object Completed : ExitCase<Nothing>()
+
+  object Cancelled : ExitCase<Nothing>()
+
+  data class Error<out E>(val e: E) : ExitCase<E>()
+}
+
+fun <E> Either<E, *>.toExitCase() =
+  fold(::Error) { ExitCase.Completed }
+
+/**
+ * Extension of MonadError exposing the [[bracket]] operation, a generalized abstracted pattern of safe resource
+ * acquisition and release in the face of errors or interruption.
+ *
+ * @define The functions receiver here (Kind<F, A>) would stand for the "acquireParam", and stands for an action that
+ * "acquires" some expensive resource, that needs to be used and then discarded.
+ *
+ * @define use is the action that uses the newly allocated resource and that will provide the final result.
+ */
+interface Bracket<F, E> : MonadError<F, E> {
+
+  /**
+   * A generalized version of [[bracket]] which uses [[ExitCase]] to distinguish between different exit cases when
+   * releasing the acquired resource.
+   *
+   * @param release is the action supposed to release the allocated resource after `use` is done, by observing and
+   * acting on its exit condition.
+   */
+  fun <A, B> Kind<F, A>.bracketCase(use: (A) -> Kind<F, B>, release: (A, ExitCase<E>) -> Kind<F, Unit>): Kind<F, B>
+
+  /**
+   * Meant for specifying tasks with safe resource acquisition and release in the face of errors and interruption.
+   * It would be the the equivalent of `try/catch/finally` statements in mainstream imperative languages for resource
+   * acquisition and release.
+   *
+   * @param release is the action that's supposed to release the allocated resource after `use` is done, irregardless
+   * of its exit condition.
+   */
+  fun <A, B> Kind<F, A>.bracket(use: (A) -> Kind<F, B>, release: (A) -> Kind<F, Unit>): Kind<F, B> =
```

Comment:
For readability purposes, I'd move the use to the last parameter. That way we have the execution always right-aligned.

`bla.bracket({ a -> callNetwork()}) { a.cancel() }`

vs 

`bla.bracket({ a.cancel() }) { a -> callNetwork()}`

Reply:
Turns out I kept the same ordering found in cats mostly because of the bracket semantics: You take the resource, use it and then release it. But your point is also good though.

Reply:
I prefer the release/use style, for the reasons above. We can ask in the channel.

Reply:
Reaffirming myself on the function order here :D

Reply:
Yeye, don't think I forgot, that'll be done before merging.